File: frontend/src/api/feedback.ts
import axios from 'axios';
import type {
  Feedback,
  FeedbackCreate,
  FeedbackUpdate,
  FeedbackListResponse,
  FeedbackStats,
  AnalysisResult,
  SmartFeedbackCreate,
} from '../types/feedback';
import type { Category, Sentiment, Priority, Status } from '../types/feedback';

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1',
  headers: {
    'Content-Type': 'application/json',
  },
});

api.interceptors.request.use((config) => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

api.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem('token');
    }
    return Promise.reject(error);
  }
);

export interface FeedbackFilters {
  category?: Category;
  sentiment?: Sentiment;
  priority?: Priority;
  status?: Status;
}

export const createFeedback = async (data: FeedbackCreate): Promise<Feedback> => {
  const response = await api.post<Feedback>('/feedback', data);
  return response.data;
};

export const getFeedbacks = async (
  page: number = 1,
  size: number = 10,
  filters?: FeedbackFilters
): Promise<FeedbackListResponse> => {
  const params: Record<string, string | number> = { page, size };
  if (filters?.category) params.category = filters.category;
  if (filters?.sentiment) params.sentiment = filters.sentiment;
  if (filters?.priority) params.priority = filters.priority;
  if (filters?.status) params.status = filters.status;

  const response = await api.get<FeedbackListResponse>('/feedback', { params });
  return response.data;
};

export const getFeedback = async (id: string): Promise<Feedback> => {
  const response = await api.get<Feedback>(`/feedback/${id}`);
  return response.data;
};

export const updateFeedback = async (
  id: string,
  data: FeedbackUpdate
): Promise<Feedback> => {
  const response = await api.put<Feedback>(`/feedback/${id}`, data);
  return response.data;
};

export const deleteFeedback = async (id: string): Promise<void> => {
  await api.delete(`/feedback/${id}`);
};

export interface StatsFilters {
  start_date?: string;
  end_date?: string;
  category?: string;
  priority?: string;
  sentiment?: string;
}

export const getStats = async (filters?: StatsFilters): Promise<FeedbackStats> => {
  const params: Record<string, string> = {};
  if (filters?.start_date) params.start_date = filters.start_date;
  if (filters?.end_date) params.end_date = filters.end_date;
  if (filters?.category) params.category = filters.category;
  if (filters?.priority) params.priority = filters.priority;
  if (filters?.sentiment) params.sentiment = filters.sentiment;

  const response = await api.get<FeedbackStats>('/feedback/stats/summary', { params });
  return response.data;
};

export const analyzeText = async (text: string, rating?: number): Promise<AnalysisResult> => {
  const response = await api.post<AnalysisResult>('/intelligence/analyze', { text, rating });
  return response.data;
};

export const analyzeExistingFeedback = async (id: string): Promise<AnalysisResult> => {
  const response = await api.post<AnalysisResult>(`/intelligence/analyze/${id}`);
  return response.data;
};

export const smartSubmitFeedback = async (data: SmartFeedbackCreate): Promise<Feedback> => {
  const response = await api.post<Feedback>('/intelligence/smart-submit', data);
  return response.data;
};

export default api;

File: backend/app/api/v1/feedback.py
"""
Feedback CRUD API routes.
"""

import logging
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.feedback import (
    FeedbackCreate,
    FeedbackListResponse,
    FeedbackResponse,
    FeedbackStats,
    FeedbackUpdate,
)
from app.services.feedback import FeedbackService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/feedback", tags=["Feedback"])


@router.get(
    "/stats/summary",
    response_model=FeedbackStats,
    summary="Get feedback statistics",
)
async def get_stats(
    start_date: str | None = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: str | None = Query(None, description="End date (YYYY-MM-DD)"),
    category: str | None = Query(None, description="Filter by category"),
    priority: str | None = Query(None, description="Filter by priority"),
    sentiment: str | None = Query(None, description="Filter by sentiment"),
    db: AsyncSession = Depends(get_db),
):
    """Get aggregated statistics for the dashboard with optional query filters."""
    from datetime import datetime, time
    
    start_dt = None
    if start_date:
        try:
            start_dt = datetime.combine(datetime.strptime(start_date, "%Y-%m-%d").date(), time.min)
        except ValueError:
            pass
            
    end_dt = None
    if end_date:
        try:
            end_dt = datetime.combine(datetime.strptime(end_date, "%Y-%m-%d").date(), time.max)
        except ValueError:
            pass

    return await FeedbackService.get_feedback_stats(
        db,
        start_date=start_dt,
        end_date=end_dt,
        category=category,
        priority=priority,
        sentiment=sentiment,
    )


@router.post(
    "",
    response_model=FeedbackResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create feedback",
)
async def create_feedback(
    data: FeedbackCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new feedback entry."""
    return await FeedbackService.create_feedback(db, data)


@router.get(
    "",
    response_model=FeedbackListResponse,
    summary="List feedback",
)
async def list_feedback(
    page: int = Query(1, ge=1, description="Page number"),
    size: int = Query(10, ge=1, le=100, description="Items per page"),
    category: str | None = Query(None, description="Filter by category"),
    status_filter: str | None = Query(None, alias="status", description="Filter by status"),
    priority: str | None = Query(None, description="Filter by priority"),
    sentiment: str | None = Query(None, description="Filter by sentiment"),
    db: AsyncSession = Depends(get_db),
):
    """Get a paginated list of feedback with optional filters."""
    return await FeedbackService.get_feedbacks(
        db, page=page, size=size,
        category=category, status=status_filter,
        priority=priority, sentiment=sentiment,
    )


@router.get(
    "/{feedback_id}",
    response_model=FeedbackResponse,
    summary="Get feedback by ID",
)
async def get_feedback(
    feedback_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get a single feedback entry by its UUID."""
    return await FeedbackService.get_feedback(db, feedback_id)


@router.patch(
    "/{feedback_id}",
    response_model=FeedbackResponse,
    summary="Update feedback",
)
async def update_feedback(
    feedback_id: UUID,
    data: FeedbackUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update an existing feedback entry. Only provided fields are updated."""
    return await FeedbackService.update_feedback(db, feedback_id, data)


@router.delete(
    "/{feedback_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete feedback",
)
async def delete_feedback(
    feedback_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Delete a feedback entry by its UUID."""
    await FeedbackService.delete_feedback(db, feedback_id)
